### PR TITLE
Allow use of Junebug as a message gateway.

### PR DESCRIPTION
**Pull request by smn**

### CODE REVIEW

**Comment by smn** [comment]
@imsickofmaps please let me know if I'm way off target here. I'll be adding tests for the factory tonight or tomorrow.

**Comment by imsickofmaps** [comment]
@smn looks on target to me. The issue that will come up next is the outbound message payload needs to be different for junebug, so there needs to be some mapping done. http://junebug.readthedocs.io/en/latest/http_api.html#messages 

I avoided this work to begin with due to Junebug not being everywhere yet, with the intention of doing Junebug only (to take advantage of Junebugs better API's) once it was universal.

**Comment by smn** [comment]
@imsickofmaps you mentioned something about event handling needing to be changed as acks & nacks are slightly different syntaxes for Junebug.

Is that required for this? We're simply sending a message out via HTTP, we're not expecting anything returned right?

**Comment by smn** [comment]
parking this for now

**Comment by rudigiesler** [comment]
@smn Ready for review.

File: message_sender/tests.py [hunk 1]
```diff
@@ -452,6 +455,139 @@ def test_event_nack_last(self):
         #     self.check_logs("Metric: 'vumimessage.maxretries' [sum] -> 1"))
 
 
+@override_settings(MESSAGE_BACKEND='junebug',
+                   ROOT_URLCONF='message_sender.urls')
+class TestJunebugMessagesAPI(AuthenticatedAPITestCase):
+    def setUp(self, *args, **kwargs):
+        reload(sys.modules[settings.ROOT_URLCONF])
```
Comment (smn) [inline message_sender/tests.py:463]: This seems to break in py34 according to travis

File: message_sender/factory.py [hunk 2]
```diff
@@ -54,11 +54,14 @@ def _raw_send(self, py_data):
...
-        data = json.dumps(py_data)
+        data = json.dumps(data)
         r = self.session.post(self.api_url, auth=self.auth,
                               data=data, headers=headers)
         r.raise_for_status()
-        return r.json()
+        res = r.json()
+        return {
+            "message_id": res.get('result', {}).get('id')
+        }
```
Comment (smn) [inline message_sender/factory.py]: Is this the only relevant key we need here? No need to provide any more info (which would probably mean needing to add mappings of keys)

File: message_sender/tests.py (hunk 1 above)
Comment (smn) [reply message_sender/tests.py:463]: And why are we needing to reload this?

**Review by smn — COMMENTED** [review]
Looks good, just a question about the fields required for the JunebugApiSender's response and the need for reload()

File: message_sender/tests.py (hunk 1 above)
Comment (rudigiesler) [reply message_sender/tests.py:463]: In our urls file, we are setting the route for `/api/v1/events` according to `EventListenerFactory.create()`, so if we just override the setting, it still points to the old vumi event view, not the new Junebug one, so we have to reload the URLs for it to pick up the changed setting and point to the right view.

File: message_sender/factory.py (hunk 2 above)
Comment (rudigiesler) [reply message_sender/factory.py]: This is the only key that is being used at the moment. (the only place in the codebase that it's being used is in the task, and the task just uses the `message_id` field).

Comment (rudigiesler) [reply message_sender/factory.py]: Actually, I think I'll change this to return the whole result dictionary, that way it is futureproof.

Comment (rudigiesler) [reply message_sender/factory.py]: (it won't be perfectly mapped, but at least the information is all returned.)

**Comment by smn** [comment]
:+1:

**Pull request merged**